nes/src/cpu/instruction/mod.rs

```diff
@@ -14,56 +15,20 @@ macro_rules! nes_asm {
 }
 
 mod and;
-mod jmp;
-mod lda;
-mod nop;
 
-#[macro_export]
-macro_rules! instruction_match {
-    ($op:ident, $fn:ident, $($a:ident),*) => {
-        match $op {
-            <and::Immediate as Execute>::OPCODE => <and::Immediate as Execute>::$fn($($a),*),
-            <and::ZeroPage as Execute>::OPCODE => <and::ZeroPage as Execute>::$fn($($a),*),
-            <and::ZeroPageX as Execute>::OPCODE => <and::ZeroPageX as Execute>::$fn($($a),*),
-            <and::Absolute as Execute>::OPCODE => <and::Absolute as Execute>::$fn($($a),*),
-            <and::AbsoluteX as Execute>::OPCODE => <and::AbsoluteX as Execute>::$fn($($a),*),
-            <and::AbsoluteY as Execute>::OPCODE => <and::AbsoluteY as Execute>::$fn($($a),*),
-            <and::IndirectX as Execute>::OPCODE => <and::IndirectX as Execute>::$fn($($a),*),
-            <and::IndirectY as Execute>::OPCODE => <and::IndirectY as Execute>::$fn($($a),*),
-
-            <jmp::Absolute as Execute>::OPCODE => <jmp::Absolute as Execute>::$fn($($a),*),
-            <jmp::Indirect as Execute>::OPCODE => <jmp::Indirect as Execute>::$fn($($a),*),
-
-            <lda::Immediate as Execute>::OPCODE => <lda::Immediate as Execute>::$fn($($a),*),
-            <lda::ZeroPage as Execute>::OPCODE => <lda::ZeroPage as Execute>::$fn($($a),*),
-            <lda::ZeroPageX as Execute>::OPCODE => <lda::ZeroPageX as Execute>::$fn($($a),*),
-            <lda::Absolute as Execute>::OPCODE => <lda::Absolute as Execute>::$fn($($a),*),
-            <lda::AbsoluteX as Execute>::OPCODE => <lda::AbsoluteX as Execute>::$fn($($a),*),
-            <lda::AbsoluteY as Execute>::OPCODE => <lda::AbsoluteY as Execute>::$fn($($a),*),
-            <lda::IndirectX as Execute>::OPCODE => <lda::IndirectX as Execute>::$fn($($a),*),
-            <lda::IndirectY as Execute>::OPCODE => <lda::IndirectY as Execute>::$fn($($a),*),
-
-            <nop::Implicit as Execute>::OPCODE => <nop::Implicit as Execute>::$fn($($a),*),
-
-            _ => panic!("instruction not implemented: 0x{:02X}", $op),
-        }
-    };
-    ($op:ident, $fn:ident) => (instruction_match!($op, $fn,))
-}
-
-trait Execute {
-    const OPCODE: u8;
-    const CYCLES: usize;
-    const PAGE_BOUNDARY_EXTRA_CYCLES: bool = false;
+// TODO(benjaminjt): Generate this with a macro, e.g. instruction_set![and::immediate, ...]
+mod instruction_set {
+    use super::and;
+    use super::Instruction;
 
-    fn exec<T: ReadAddr + WriteAddr>(core: &mut Core, memory: &mut T);
+    const AND_IMMEDIATE_CODE: u8 = and::IMMEDIATE.opcode;
+    const AND_ABSOLUTE_CODE: u8 = and::ABSOLUTE.opcode;
 
-    #[inline(always)]
-    fn to_instruction() -> Instruction {
-        Instruction {
-            opcode: Self::OPCODE,
-            cycles: Self::CYCLES,
-            page_boundary_extra_cycle: Self::PAGE_BOUNDARY_EXTRA_CYCLES,
+    pub fn get(opcode: u8) -> Instruction {
+        match opcode {
+            AND_IMMEDIATE_CODE => and::IMMEDIATE,
```

Comment:
I don't know why, but this doesn't work:
```rust
match opcode {
  and::IMMEDIATE.opcode => and::IMMEDIATE,
  ...
}
```

Reply:
What's the Cargo error trace?

Reply:
I think this might work ?

```rust
match opcode {
  o if o == and::IMMEDIATE.opcode => and::IMMEDIATE,
  ...
}
```